apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableReadRowsRequestManager.cs

```diff
@@ -0,0 +1,134 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Protobuf.Collections;
+using System.Diagnostics;
+using Google.Protobuf;
+using System.Linq;
+using System;
+using System.Collections.Generic;
+
+namespace Google.Cloud.Bigtable.V2
+{
+    /// <summary>
+    /// Keeps track of Rows returned from a readRows RPC for information relevant to resuming the RPC
+    /// after temporary problems.
+    /// </summary>
+    internal class BigtableReadRowsRequestManager
+    {
+        // Member variables from the constructor.
+        private readonly ReadRowsRequest _originalRequest;
+
+        // The number of rows read so far.
+        internal long RowCount { get; private set; }
+
+        /// <summary>
+        /// Constructor for <see cref="BigtableReadRowsRequestManager"/>.
+        /// </summary>
+        /// <param name="originalRequest">
+        /// Original ReadRowsRequest containing all of the parameters of the API call.
+        /// </param>
+        internal BigtableReadRowsRequestManager(ReadRowsRequest originalRequest) => _originalRequest = originalRequest;
+
+        /// <summary>
+        /// Update last found Rowkey
+        /// </summary>
+        internal BigtableByteString LastFoundKey { get; set; }
+
+        /// <summary>
+        /// Update amount of rows read so far
+        /// </summary>
+        internal void IncrementRowCount(int count) => RowCount += count;
+
+        /// <summary>
+        /// Builds and returns updated subrequest that excludes all rowKeys that have already been found.
+        /// </summary>
+        internal ReadRowsRequest BuildUpdatedRequest()
+        {
+            ReadRowsRequest newReadRowsRequest = new ReadRowsRequest
+            {
+                // Transfer TableName
+                TableNameAsTableName = _originalRequest.TableNameAsTableName,
+                // Transfer RowFilter.
+                Filter = _originalRequest.Filter,
+                // Remove received rowKeys form Rows.
+                Rows = FilterRows()
+            };
+
+            // If the row limit is set, update it.
+            if (_originalRequest.RowsLimit != 0)
+            {
+                long rowsRemaining = _originalRequest.RowsLimit - RowCount;
+                Debug.Assert(rowsRemaining > 0, "The remaining number of rows must be greater than 0.");
+                newReadRowsRequest.RowsLimit = Math.Max(1, rowsRemaining);
+            }
+
+            return newReadRowsRequest;
+        }
+
+        /// <summary>
+        /// Removes received rows from the RowSet.
+        /// </summary>
+        /// <returns>
+        /// New RowSet with not received rows.</returns>
+        private RowSet FilterRows()
+        {
+            RowSet originalRows = _originalRequest.Rows;
+
+            if (LastFoundKey.Value.IsEmpty)
```

Comment:
I see this used a bunch here. I think it probably makes sense to add an IsEmpty delegating property on `BigtableByteString`. I can add this and update the code here after it's merged.

Reply:
Currently it could be done as I did,
1. Get the underlying byte string and check the value using `IsEmpty` property.
or
2. Get the underlying byte string and check the value against the `null` -> 
`LastFoundKey.Value == null`

Reply:
There's also a 3rd way: `LastFoundKey.Lenght == 0`. However, that said, I'm adding `IsEmpty` in #1822, so when that's merged I'll fix up these references.